WT-7329 Add hook capability to Python testing.

ddanderson opened the pull request.

ddanderson commented:
@keitharnoldsmith and @boboeee, could you please review this?  There's no big rush, nobody's waiting on it.  The idea is that going forward we'll create something like `hook_tiered.py` that makes a number of hooks so that tiered storage is enabled on tests and that `session.create` calls are modified to do tiering.  This will give us a lot of tests for tiering real quickly. This is a more general facility than we'll need right now (as you can hook on arbitrary methods), but it seems like it will be useful in the future.

No PR testing is going to run with any hooks right now, but if/when we get a body of tests that pass with the tiered hook, we may set up a evergreen test to run it.  Output looks something like this (running with demo=2 injects an extra `drop` and `create` from what is in the original test).
```
python3 ../test/suite/run.py --hook demo=2 base01
>> SETUP HOOKS RUN, SET TO <built-in function Session_create>
Setting up hook on <module 'swig_wiredtiger' from '/Users/dda/.dotfiles/wt/git/wt-7329-python-test-hooks/build_posix/lang/python/wiredtiger/swig_wiredtiger.py'>.wiredtiger_open
Setting up hooks list in <module 'swig_wiredtiger' from '/Users/dda/.dotfiles/wt/git/wt-7329-python-test-hooks/build_posix/lang/python/wiredtiger/swig_wiredtiger.py'>._wiredtiger_open_hooks
Setting up hooks list in <module 'swig_wiredtiger' from '/Users/dda/.dotfiles/wt/git/wt-7329-python-test-hooks/build_posix/lang/python/wiredtiger/swig_wiredtiger.py'>._wiredtiger_open_hooks
Setting up hook on <class 'swig_wiredtiger.Session'>.create
Setting up hooks list in <class 'swig_wiredtiger.Session'>._create_hooks
Setting up hook on <class 'swig_wiredtiger.Session'>.open_cursor
Setting up hooks list in <class 'swig_wiredtiger.Session'>._open_cursor_hooks
Filtering: <unittest.suite.TestSuite tests=[<test_base01.test_base01 testMethod=test_empty>, <test_base01.test_base01 testMethod=test_error>, <test_base01.test_base01 testMethod=test_insert>]>
>>> wiredtiger_open, adding cache_size
>>> wiredtiger_open(('.', 'create,error_prefix="test_base01.test_base01.test_empty",,,,cache_size=500M,,,')) returned <module 'swig_wiredtiger' from '/Users/dda/.dotfiles/wt/git/wt-7329-python-test-hooks/build_posix/lang/python/wiredtiger/swig_wiredtiger.py'>
>>> session.create(table:test_base01a.wt,key_format=S,value_format=S,allocation_size=512,internal_page_max=16384,leaf_page_max=131072), session is <swig_wiredtiger.Session; proxy of <Swig Object of type 'WT_SESSION *' at 0x7fba14b42030> >
>>> session.drop(table:test_base01a.wt)
>>> session.create(table:test_base01a.wt,key_format=S,value_format=S,allocation_size=512,internal_page_max=16384,leaf_page_max=131072)
>>> session.open_cursor(('table:test_base01a.wt', None, None)) returned <swig_wiredtiger.Session; proxy of <Swig Object of type 'WT_SESSION *' at 0x7fba14b42030> >, session is <swig_wiredtiger.Cursor; proxy of <Swig Object of type 'WT_CURSOR *' at 0x7fba14b42090> >
.>>> wiredtiger_open, adding cache_size
>>> wiredtiger_open(('.', 'create,error_prefix="test_base01.test_base01.test_error",,,,cache_size=500M,,,')) returned <module 'swig_wiredtiger' from '/Users/dda/.dotfiles/wt/git/wt-7329-python-test-hooks/build_posix/lang/python/wiredtiger/swig_wiredtiger.py'>
>>> session.create(table:test_base01a.wt,expect_this_error,okay?), session is <swig_wiredtiger.Session; proxy of <Swig Object of type 'WT_SESSION *' at 0x7fba14b42030> >
.>>> wiredtiger_open, adding cache_size
>>> wiredtiger_open(('.', 'create,error_prefix="test_base01.test_base01.test_insert",,,,cache_size=500M,,,')) returned <module 'swig_wiredtiger' from '/Users/dda/.dotfiles/wt/git/wt-7329-python-test-hooks/build_posix/lang/python/wiredtiger/swig_wiredtiger.py'>
>>> session.create(table:test_base01b.wt,key_format=S,value_format=S,allocation_size=512,internal_page_max=16384,leaf_page_max=131072), session is <swig_wiredtiger.Session; proxy of <Swig Object of type 'WT_SESSION *' at 0x7fba14b42030> >
>>> session.drop(table:test_base01b.wt)
>>> session.create(table:test_base01b.wt,key_format=S,value_format=S,allocation_size=512,internal_page_max=16384,leaf_page_max=131072)
>>> session.open_cursor(('table:test_base01b.wt', None, None)) returned <swig_wiredtiger.Session; proxy of <Swig Object of type 'WT_SESSION *' at 0x7fba14b42030> >, session is <swig_wiredtiger.Cursor; proxy of <Swig Object of type 'WT_CURSOR *' at 0x7fba14b420f0> >
>>> session.open_cursor(('table:test_base01b.wt', None, None)) returned <swig_wiredtiger.Session; proxy of <Swig Object of type 'WT_SESSION *' at 0x7fba14b42030> >, session is <swig_wiredtiger.Cursor; proxy of <Swig Object of type 'WT_CURSOR *' at 0x7fba14b42090> >
.
----------------------------------------------------------------------
Ran 3 tests in 2.532s

OK
```

jiechenbo commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,216 @@
+        self.hookmgr.add_hook(wiredtiger.Cursor, method_name, hook_type, hook_func)
+
+    # Allows lookup of the original version of functions
+    def get_wiredtiger_function(self, method_name):
```
Would merging all of these into one getter and setter function be more easier to understand? Applying something like adding **object_name** into the method, and grabbing the object via wiredtiger[object_name].

ddanderson replied:
I took a stab at this - not quite sure your vision of this was.  Here's what I came up with:
```
            orig_session_create = self.Session['create']
            self.Session['open_cursor'] = (wthooks.HOOK_NOTIFY, session_open_cursor_notify)
```
I'm using `self.Session` rather than `Session`, because I was a little uncomfortable overriding the setitem/getitem on the `Connection/Session/Cursor` class or `wiredtiger` module.  `self.Session` is a proxy object that gets the same thing done.

The assignment needs two arguments, hence a tuple.

It's _slightly_ unintuitive in that an "assignment" adds a hook, but doesn't override any existing hook, unlike regular assignment.  We could also do something like:
```
       self.Session['open_cursor'].append(wthooks.HOOK_NOTIFY, session_open_cursor_notify)
```

jiechenbo replied:
Sorry for the misunderstanding, my initial vision was to combine the multiple functions through something like:
**
def get_connection_function(self, method_name, object_name=None):
    if object_name != None and hasattr(wiredtiger, object_name):
        return self.hookmgr.get_function(wiredtiger[object_name], method_name)
    else:
        return self.hookmgr.get_function(wiredtiger, method_name)
** 
Although this version works well too, only issue might be the confusion when assigning more functions to a particular module e.g. wiredtiger, as it actually represents an **additem**, and adding the complexity of the proxy class.

jiechenbo commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,216 @@
+        return tests
+
+# Hooks must derive from this class
+class WiredTigerHook(object):
```
Would converting this to an abstract base class be more beneficial? Enforcing that WiredTigerHook needs to be derived from another class

ddanderson replied:
Thanks - I did this.  My first time using "abc", so please check my work.

jiechenbo commented on test/suite/hook_demo.py:
```diff
@@ -0,0 +1,130 @@
+# Every hook file must have one or more classes descended from WiredTigerHook
+# This is where the hook functions are 'hooked' to API methods.
+class DemoHook(wthooks.WiredTigerHook):
+    def __init__(self, arg=0):
```
What happens if an hook object, decides to have two functions that use arg? Is the consensus that 2nd function that uses args needs to start from an offset? Might cause a lot of confusion down a complex hook object

ddanderson replied:
It's totally legal to have two functions that use arg.  It's up to `DemoHook` coordinate things if it wants something complex.  For example, `DemoHook` could accept an argument like` --hook demo="debug,no_open_cursor_hook,session.create.arg=27,wiredtiger_open.arg=17"`.  Which might mean:
- pass an extra debug arg to all functions
- don't set up the `open_cursor` hook.
- `session.create` gets an extra argument of 27
etc.

You get the idea, it can be arbitrarily complex, and the DemoHook needs to know what to do.  Though I expect if we get to doing stuff like this, we'll make helper functions and evolve as needed.  Let's see where it goes....

jiechenbo reviewed (changes requested):
Hey @ddanderson, this ticket reaches amazing usage of python coding. Really interesting and cool stuff here! I have some comments, regarding some of the knowledge here. This is just initial review. I have still yet to look at how notify, replace and args work in detail. But overall some interesting concepts done here.

sueloverso commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,216 @@
+            notifies.append(hook_func)
+        elif hook_type == HOOK_REPLACE:
+            replace_func = hook_func
+        elif hook_type == HOOK_ARGS:
```
Alphabetize this to the top.

ddanderson replied:
Fixed.

sueloverso commented on test/suite/run.py:
```diff
@@ -119,6 +119,7 @@ def usage():
                                  be run without executing any.\n\
   -g      | --gdb                all subprocesses (like calls to wt) use gdb\n\
   -h      | --help               show this message\n\
+          | --hook name[=arg]    set up hooks from hook_name.py, with optional arg\n\
```
I don't see `hook_name.py`. Oh, I get it now. Would it be clearer as `hook_<name>.py`?

ddanderson replied:
Fixed.

sueloverso commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,216 @@
+        names_seen = []
+        for name in hooknames:
+            # The hooks are indicated as "somename=arg" or simply "somename".
+            # hook_somename.py will be imported, and initialized with the arg.
```
I think this needs to be repeated or more prominently displayed at the top of this file so that someone looking at adding a hook can find it easily in here.

ddanderson replied:
I've added a big comment section in `wthooks.py` that also mentions this, among other internal details.

sueloverso reviewed (approved):
I have a few minor comments. LGTM

ddanderson commented:
@keitharnoldsmith, I know you're already experimenting with this - I've changed the interface slightly according to @jiechenbo's ideas, I recommend you don't merge into your experiments until we nail down the specifics.

I've also added more comments, hopefully it will be easier to understand how it works.

jiechenbo commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,256 @@
+        #    wiredtiger.Connection._XXXX_hooks = []
+        #    wiredtiger.Connection._XXXX_orig = wiredtiger.Connection.XXXX
+        #    wiredtiger.Connection.XXXX = lambda self, *args: hooked_function(self, 'XXXX', *args)
+        hooks_name = '_' + method_name + '_hooks'
```
would like **hooks_name** to infer that its a list, maybe something like hooks_list here?

ddanderson replied:
Because of your previous comment, the hooks is no longer a list, it's a `WiredTigerHookInfo`, so I renamed this `hook_info_name`.

jiechenbo commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,256 @@
+            orig_func = getattr(clazz, method_name)
+        return orig_func
+
+    def filter_tests(self, tests):
```
Does this imply that multiple hooks will always create a smaller or equal test suite? What happens if you want a hook to specifically run under a specific test suite, and another hook to run another different test suite? That would mean no tests would be run, let me know if that's the intended use case

ddanderson replied:
Yes, multiple hooks will run a smaller or equal test suite. It's a very good point.  I think this filter was less intended to be used as "only run the cursor tests with this hook", but more like, "don't run the backup tests with this hook, there's something fundamentally broken about that combination".   Similar to how we "skip" tests today, quite targeted.  So, yes it's possible that combining multiple hooks might get you to empty set, I think the filters should be cautious about removing too much.  Especially if we have multiple hooks in mind.  For at least a while, I'm suspecting that multiple hooking will either be rare or else someone may create a few hooks at the same time specially expecting them to be used together.

ddanderson commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,256 @@
+
+# hooked_function -
+# A helper function for the hook manager.
+def hooked_function(self, orig_func, hooks_name, *args):
```
By passing in the function, rather than the string, we avoid doing a `hasattr()` or `getattr()` on every call to the function.  Some of the functions may be very high usage (like hooks on any of the `Cursor` methods, especially).

jiechenbo replied:
oh yeah, that makes alot of sense!

ddanderson commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,256 @@
+        self.hookmgr.add_hook(self.clazz, name, hooktype, fcn)
+
+# Hooks must derive from this class
+class WiredTigerHookCreator(metaclass = abc.ABCMeta):
```
Done.

ddanderson commented on test/suite/wthooks.py:
```diff
@@ -0,0 +1,256 @@
+    #
+    # We only walk through the hook list once, and process the config
+    # hooks while we're doing that, and copy any other hooks needed.
+    for hook_type,hook_func in hook_func_list:
```
I modified this a little bit.  There is no longer a hook_func_list, but a hook_info object.  That object has three attributes that correspond to the two lists and the replacement function.

ddanderson commented:
@jiechenbo, hopefully I addressed your comments.  I did not change the interface for the hook creator, I don't want to change it again until we settle on a good approach.  Currently it looks like:
```
        orig_session_create = self.Session['create']     # gets original function
        self.wiredtiger['wiredtiger_open'] = (wthooks.HOOK_ARGS, wiredtiger_open_args)
        self.wiredtiger['wiredtiger_open'] = (wthooks.HOOK_NOTIFY, wiredtiger_open_notify)
        self.Session['create'] = (wthooks.HOOK_REPLACE, lambda s, uri, config:
          session_create_replace(self.arg, orig_session_create, s, uri, config))
        self.Session['open_cursor'] = (wthooks.HOOK_NOTIFY, session_open_cursor_notify)
```
Perhaps in the future, if we think it's worth the effort, I think we could do this:
```
        orig_session_create = self.Session.create.original_function
        self.wiredtiger.wiredtiger_open.args_functions += wiredtiger_open_args
        self.wiredtiger.wiredtiger_open.notify_functions += wiredtiger_open_notify
        self.Session.create.replace_functions += lambda s, uri, config:
          session_create_replace(self.arg, orig_session_create, s, uri, config)
        self.Session.open_cursor.notify_functions += session_open_cursor_notify
```
Or `+=` could be `append(...)` instead.  Anyway, I don't think it will matter in the near term, the number of hook_*.py files and functions needing to be hooked will likely be small.

jiechenbo reviewed (approved):
This feature looks great! Thank you for allowing me to review this. Some really cool concepts here. Regarding the design, I am fine with both. LGTM!

ddanderson commented:
@keitharnoldsmith, I didn't know if you had any review comments for this? I think we've stabilized on a design for the interface, which will require some small changes (no changes to your functions, just how you "hook" them).

keitharnoldsmith commented on test/suite/hook_demo.py:
```diff
@@ -0,0 +1,130 @@
+#  - notify on session.open_cursor
+#  - intercept the session.create call
+#
+#   With N == 0, the session.create call merely reports its arguments;
```
And, with N == 0, it calls the original create.

ddanderson replied:
Fixed.

keitharnoldsmith reviewed (approved):
LGTM!

One minor note about the explanatory comments.

I learned a bunch about Python reading through this, but don't have a lot of comments about the actual coding.

From using the hooks to run some tests, I can say that the code, interface, and comments all make it quite easy to use.
